portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -22,22 +25,47 @@
 
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
I think you're really close here with sending back JSON. Here's what I was thinking at a highish level:

To start, you'll need to figure out what data you want to send back and forth between the client and the server (aka, you'll need to define your API). It seems like you want to send the user's email and a URL to login or to logout. It looks like you've already got that defined in your `Login` class. To make it really clear what that class is, I might name it something like `LoginApi` or `AuthApi`.

Next in doGet, you've got pretty much everything you need already. Instead of returning the URL and username embedded in the HTML, you can do something like

```java
LoginApi api;
if (userService.isUserLoggedIn()) {
    String userEmail = userService.getCurrentUser().getEmail();
    String urlToRedirectToAfterUserLogsOut = "/";
    String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
    api = new LoginApi(userEmail, logoutUrl);
} else {
    String urlToRedirectToAfterUserLogsIn = "/";
    String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
    api = new LoginApi(null, loginUrl);  // Not sure if you can pass null here, if you can't then pass "" instead.
}
Gson gson = new Gson();
response.setContentType("application/json");
response.getWriter().println(gson.toJson(api));
```

Then you'll need to make some changes in the front end. They should make things a bit simpler for you. For example, down below where you have `const username = user.split("<p");` you can instead use `const username = user["useremail"]`. (As an aside, you might want to make that `userEmail` just so it's easier to read).

For the `loginStatus`, you can change that to be something like

```javascript
// If the user's email is null or empty, then they're not logged in.
let loginStatus;
if (user["useremail"] !== null && user["useremail"] !== "") {
    // We format using JavaScript template literals https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals
    loginStatus = `Hello stranger please sign in to leave a comment. <a href="${user['url']}">Login here</a>`;
} else {
    loginStatus = `Hello ${user['useremail']}, <a href="${user['url']}">Logout here</a>`;
}
```
If you want to do additional formatting, that's also easier now since you can use the browser's DOM API to construct things instead of needing to `"<p><a href=\"" + stuff + ...`.